### PR TITLE
BREAKING: Enable ARM-based ACA-Py artifacts by default by removing BBS+ Signatures as a default inclusion

### PR DESCRIPTION
**Pull request by amanji**
In order to have ACA-Py artifacts be cross-architecture compatible out of the box, this PR removes the `bbs` optional dependency from the base builds of ACA-Py. `bbs` doesn't currently support ARM architecture (aarch64 or arm64) therefore BBS+ is not available for ACA-Py on ARM-based devices (e.g. Apple M1 Macs). BBS+ can still be used in ACA-Py on non-ARM-based devices and there is documentation about how to do that however, it must be explicitly added in.

See the issues #1695 and #3104 and others concerned with using ACA-Py on an Apple M1 Mac out of the box.

For base ACA-Py builds BBS+ unit tests are skipped and integration tests that require BBS+ are marked as such. In order for GHA to pass these integration tests will need to be skipped with `-t ~@BBS`.

An additional publish step has been added to generate ACA-Py Docker images with BBS+. These are tagged as `aries-cloudagent-bbs`.

### CODE REVIEW

**Comment by amanji** [comment]
Related to: #3104

**Review by dbluhm — APPROVED** [review]

File: docker/Dockerfile.indy
Comment (amanji) [inline docker/Dockerfile.indy:1]: @dbluhm I assume this is ok to remove? Do you know by any chance if it is required anymore?

Comment (dbluhm) [reply docker/Dockerfile.indy:1]: No longer required -- all the indy code was removed in a previous PR; looks like this one was just overlooked

**Review by jamshale — COMMENTED** [review]
Looks good :+1: I'm just curious if we want to skip these tests or not or if we could install bbs for the workflows and still test them?

**Comment by amanji** [comment]
Would it make sense for me to add a follow up PR for that? We can add a specific workflow that installs bbs and tests them. I just don't want 1.0.0 being held up

**Comment by jamshale** [comment]
> Would it make sense for me to add a follow up PR for that? We can add a specific workflow that installs bbs and tests them. I just don't want 1.0.0 being held up

Ya. I think that should be fine.

**Review by jamshale — APPROVED** [review]

**Comment by swcurran** [comment]
1.0.0 can’t go out without the documentation, so might as well include it in this one.

I’ve reframed the purpose of this in the title and description. I think we need to continue to refine that.

**Review by jamshale — CHANGES_REQUESTED** [review]
After conversation I think we need to think about this a bit more before including it in 1.0.0

File: demo/features/0453-issue-credential.feature [hunk 1]
```diff
@@ -125,7 +125,7 @@ Feature: RFC 0453 Aries agent issue credential
        | --public-did --multitenant             | --multitenant             | driverslicense | Data_DL_NormalizedValues |
 
 
-  @T003.1-RFC0453
+  @T003.1-RFC0453 @BBS
```
Comment (swcurran) [inline demo/features/0453-issue-credential.feature:128]: I think this is too big a hammer — e.g. applied at too high a level.  I notice in the `steps` that `DID_METHOD_KEY` is imported with the `BLS` ones.  That tells me that the tests might be run with both DID Key (aka Ed25519 keys) and/or with BLS ones.  I think we need to be more surgical about flagging / removing BBS tests while still leaving the DID Key tests.

**Comment by swcurran** [comment]
For example, in the [Present Proof test steps](demo/features/0453-issue-credential.feature) I notice there is only `SIG_TYPE_BLS`.  Could we add `SIG_TYPE_ED255` (which I note is not there), and replace the use of BLS with ED25519?  I really don’t like the idea of the integration tests not having any JSON-LD testing.

File: demo/features/0453-issue-credential.feature (hunk 1 above)
Comment (amanji) [reply demo/features/0453-issue-credential.feature:128]: Looking at making these more granular.

Comment (amanji) [reply demo/features/0453-issue-credential.feature]: I have removed the `@BBS` tags for the scenarios and instead expanded the underlying examples to include tests specifically for ED25519 and BLS. The BLS examples are marked with `@BBS` making the scenarios more granular. Only these will be skipped when tests are run with `-t ~@BBS` while the ED25519 tests will always run.

**Comment by amanji** [comment]
> For example, in the [Present Proof test steps](demo/features/0453-issue-credential.feature) I notice there is only `SIG_TYPE_BLS`. Could we add `SIG_TYPE_ED255` (which I note is not there), and replace the use of BLS with ED25519? I really don’t like the idea of the integration tests not having any JSON-LD testing.

I have now added test examples for ED25519 along with the BLS examples

**Comment by swcurran** [comment]
Is there any way to fix the security — running as root in building in the image?

File: docs/deploying/BBSSignatures.md [hunk 2]
```diff
@@ -0,0 +1,27 @@
+# BBS+ Signatures Support
+
+ACA-Py includes support for BBS+ Signatures out of the box through an optional dependency, however the implementation for it is out of date and does not currently support ARM (aarch64 or arm64) architecture (e.g. Apple devices with M1 chips or above).
```
Comment (swcurran) [inline docs/deploying/BBSSignatures.md:3]: How about updating this paragraph to the following to better describe what we are doing:

> ACA-Py has supported BBS Signatures for some time. However, the dependency that is used does not support the ARM architecture, and its inclusion in the default ACA-Py artifacts mean that developers using MacOS M1 hardware cannot run ACA-Py “out-of-the-box”.  We feel that providing a better developer experience by supporting M1 hardware is more important than BBS Signature support at this time.  As such, we have removed the BBS dependency from the base ACA-Py artifacts, and made it an add on, that those using ACA-Py with BBS must take extra steps to build their own artifacts. This file describes how to do those extra steps.
>
> Regarding future support for BBS Signatures in ACA-Py. There is currently a lot of work going on in developing BBS Signature implementations and BBS-based verifiable credential standards. However, at the time of this release, there is not an obvious approach and BBS Signatures implementation to use in ACA-Py that includes M1 ARM support. As a result, we will hold off on updating the BBS Signatures support in ACA-Py until the standards and path forward clarify. In the meantime, the Maintainers of ACA-Py plan to continue to do all we can to push for newer and better ZKP-based verifiable credential standards.

**Review by swcurran — CHANGES_REQUESTED** [review]
See updated wording for the documentation.

**Comment by amanji** [comment]
> Is there any way to fix the security — running as root in building in the image?

This seems to be a common error that comes up due to using build-args in Docker.

@jamshale any thoughts on whether we can disable or fix the error?

**Comment by jamshale** [comment]
> > Is there any way to fix the security — running as root in building in the image?
> 
> This seems to be a common error that comes up due to using build-args in Docker.
> 
> @jamshale any thoughts on whether we can disable or fix the error?

My thinking on this was that because it was just the `demo` image, that it wouldn't be run in any prod deployment and that we could ignore it. Sonarcloud does have some suggestions about removing the vulnerbility https://sonarcloud.io/project/security_hotspots?id=hyperledger_aries-cloudagent-python&pullRequest=3127&issueStatuses=OPEN%2CCONFIRMED&sinceLeakPeriod=true&tab=how_to_fix if we think we should.

**Comment by swcurran** [comment]
I just don’t like merging things with flagged vulnerabilities.

**Comment by jamshale** [comment]
It's very strange. My local integration tests (ubuntu) are failing the same way as the workflow. Suddenly, just stop on the same json-ld test. No errors or anything logged. I'll try and see if I can get some more information. :thinking:

File: docs/deploying/BBSSignatures.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
+# BBS+ Signatures Support
+
+ACA-Py has supported BBS+ Signatures for some time. However, the dependency that is used (`bbs`) does not support ARM architecture, and its inclusion in the default ACA-Py artifacts mean that developers using ARM-based hardware (such as Apple M1 Macs or later) cannot run ACA-Py "out-of-the-box". We feel that providing a better developer experience by supporting ARM architecture is more important than BBS+ Signature support at this time. As such, we have removed the BBS+ dependency from the base ACA-Py artifacts, and made it an add-on, that those using ACA-Py with BBS+ must take extra steps to build their own artifacts. This file describes how to do those extra steps.
```
Comment (swcurran) [inline docs/deploying/BBSSignatures.md:3]: A couple of tweaks here, pretty please:

- “BBS” and “BBS+” are two different signature schemes.  I think we should just use “BBS” throughout, even if it is not 100% correct in all places.  I finally heard the history of where the “+” came from, and why it has gone away again — fun stuff.
- "does not support ARM architecture” -- should be “the ARM…”, same with “…supporting the ARM architecture”
- Please remove the second and third commas in the phrase: `As such, we have removed the BBS+ dependency from the base ACA-Py artifacts, and made it an add-on, that those using ACA-Py with BBS+ must take extra steps to …`

File: docs/deploying/BBSSignatures.md [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+Regarding future support for BBS+ Signatures in ACA-Py, there is currently a lot of work going on in developing implementations and BBS+-based Verifiable Credential standards. However, at the time of this release, there is not an obvious approach to an implementation to use in ACA-Py that includes ARM support. As a result, we will hold off on updating the BBS+ Signatures support in ACA-Py until the standards and path forward clarify. In the meantime, maintainers of ACA-Py plan to continue to do all we can to push for newer and better ZKP-based Verifiable Credential standards.
```
Comment (swcurran) [inline docs/deploying/BBSSignatures.md:5]: On this one — change the first comma, to a period (.) and make “There is currently…” a new sentence.

File: docs/deploying/BBSSignatures.md [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+If you require BBS+ for your deployment an optional "extended" ACA-Py image has been released (`aries-cloudagent-bbs`) that includes BBS+, with the caveat that it will very likely not install on ARM architecture.
+
+## Development and Testing
+
+### Developer setup
+
+If you are a contributor or are developing using a local build of ACA-Py and need BBS+, the easiest way to include it is to install the optional dependency `bbs` it with `poetry` (again with the caveat that it will very likely not install on ARM architecture). The `--all-extras` flag will install the `bbs` optional dependency in ACA-Py:
```
Comment (swcurran) [inline docs/deploying/BBSSignatures.md:13]: Remove the “it” after `bbs` in:

> the optional dependency `bbs` it with `poetry`

**Review by swcurran — CHANGES_REQUESTED** [review]
A couple more documentation tweaks.

**Comment by swcurran** [comment]
Running the AATH tests.  First try didn’t work, with the `pip install` failing.  I’m trying again with `bbs` removed….

I think we will need to have similar updates made to AATH as were made to the integration tests.  Of course, it’s a bit trickier, since these are multi-framework tests.   We can live with AATH failures, but we can’t live with the ACA=Py main agent not building at all.

 ```
=> [ 2/16] RUN apt-get update    && apt-get install -y git gnupg2 software-properties-common curl    && apt-key adv --keyserver keyserver.ubuntu.com --recv-keys CE7709D068DB5E88     32.0s
 => [ 3/16] RUN mkdir -p /aries-backchannels                                                                                                                                            0.3s
 => [ 4/16] WORKDIR /aries-backchannels                                                                                                                                                 0.1s
 => [ 5/16] ADD https://github.com/stedolan/jq/releases/download/jq-1.6/jq-linux64 ./jq                                                                                                 0.1s 
 => [ 6/16] RUN chmod +x ./jq                                                                                                                                                           0.3s 
 => [ 7/16] COPY python/requirements.txt python/                                                                                                                                        0.0s 
 => [ 8/16] COPY acapy/requirements-main.txt acapy/                                                                                                                                     0.1s 
 => [ 9/16] RUN pip install -r python/requirements.txt -r acapy/requirements-main.txt                                                                                                 610.3s 
 => => # INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. See https://pip.pypa.io/warnings/backtrackin 
 => => # g for guidance. If you want to abort this run, press Ctrl + C.                                                                                                                     
 => => #   Downloading attrs-22.2.0-py3-none-any.whl (60 kB)                                                                                                                                
 => => #      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.0/60.0 kB 8.3 MB/s eta 0:00:00                                                                                                    
 => => #   Downloading attrs-22.1.0-py2.py3-none-any.whl (58 kB)                                                                                                                            
 => => #      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 58.8/58.8 kB 12.4 MB/s eta 0:00:00      
```

**Comment by swcurran** [comment]
Same “not building” result with `bbs` removed from the AATH requirements file:

requirements-main.txt has:

`aries-cloudagent[indy, askar]@git+https://github.com/petridishdev/aries-cloudagent-python@chore/remove-bbs-install`

Log is stuck here after 800 seconds:

```
 => [ 6/16] RUN chmod +x ./jq                                                                                                                                                           0.3s 
 => [ 7/16] COPY python/requirements.txt python/                                                                                                                                        0.0s 
 => [ 8/16] COPY acapy/requirements-main.txt acapy/                                                                                                                                     0.0s 
 => [ 9/16] RUN pip install -r python/requirements.txt -r acapy/requirements-main.txt                                                                                                 684.1s 
 => => #   Downloading attrs-22.2.0-py3-none-any.whl (60 kB)                                                                                                                                 
 => => #      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.0/60.0 kB 5.6 MB/s eta 0:00:00                                                                                                    
 => => #   Downloading attrs-22.1.0-py2.py3-none-any.whl (58 kB)                                                                                                                            
 => => #      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 58.8/58.8 kB 13.8 MB/s eta 0:00:00                                                                                                   
 => => #   Downloading attrs-21.4.0-py2.py3-none-any.whl (60 kB)                                                                                                                            
 => => #      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.6/60.6 kB 5.9 MB/s eta 0:00:00                                                                                                    

```

Need to sort out what has to change in AATH so that ACA-Py builds again.

**Comment by amanji** [comment]
> Same “not building” result with `bbs` removed from the AATH requirements file:
> 
> requirements-main.txt has:
> 
> `aries-cloudagent[indy, askar]@git+https://github.com/petridishdev/aries-cloudagent-python@chore/remove-bbs-install`
> 
> Log is stuck here after 800 seconds:
> 
> ```
>  => [ 6/16] RUN chmod +x ./jq                                                                                                                                                           0.3s 
>  => [ 7/16] COPY python/requirements.txt python/                                                                                                                                        0.0s 
>  => [ 8/16] COPY acapy/requirements-main.txt acapy/                                                                                                                                     0.0s 
>  => [ 9/16] RUN pip install -r python/requirements.txt -r acapy/requirements-main.txt                                                                                                 684.1s 
>  => => #   Downloading attrs-22.2.0-py3-none-any.whl (60 kB)                                                                                                                                 
>  => => #      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.0/60.0 kB 5.6 MB/s eta 0:00:00                                                                                                    
>  => => #   Downloading attrs-22.1.0-py2.py3-none-any.whl (58 kB)                                                                                                                            
>  => => #      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 58.8/58.8 kB 13.8 MB/s eta 0:00:00                                                                                                   
>  => => #   Downloading attrs-21.4.0-py2.py3-none-any.whl (60 kB)                                                                                                                            
>  => => #      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.6/60.6 kB 5.9 MB/s eta 0:00:00                                                                                                    
> ```
> 
> Need to sort out what has to change in AATH so that ACA-Py builds again.

I was able to complete the build:

Screenshot 2024-07-31 080220

**Review by jamshale — APPROVED** [review]
This is looking good to me :+1:

**Review by swcurran — APPROVED** [review]
LGTM — after fixing my AATH environment to use Python 3.12, all tests pass.  I’m now comfortable with this one.  I think we are ready to merge it.

**Pull request merged**